src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -1700,6 +1705,259 @@ - (void)openCommissioningWindowWithDiscriminator:(NSNumber *)discriminator
+            AsByteSpan(mInfo.elementsSignature), AsByteSpan(mInfo.productAttestationIntermediateCertificate),
+            AsByteSpan(mInfo.deviceAttestationCertificate), AsByteSpan(mInfo.nonce),
+            static_cast<VendorId>(mInfo.basicInformationVendorID.unsignedShortValue),
+            mInfo.basicInformationProductID.unsignedShortValue);
+        mVerifier->VerifyAttestationInformation(info, &mCallback);
+    }
+
+private:
+    static void CallbackFunc(void * context, const Credentials::DeviceAttestationVerifier::AttestationInfo & info,
+        Credentials::AttestationVerificationResult result)
+    {
+        auto * self = static_cast<AttestationVerifierHelper *>(context);
+        auto * response = [[MTRDeviceAttestationResult alloc] init];
+        response.attestationInfo = self->mInfo;
+        if (result != Credentials::AttestationVerificationResult::kSuccess) {
+            MTR_LOG_ERROR("Device attestation verification failed with AttestationVerificationResult %u", to_underlying(result));
+            response.attestationSucceded = NO;
+        } else {
+            response.attestationSucceded = YES;
+        }
+
+        // We're going to delete ourselves; make sure we are not trying to use
+        // our members async.
+        auto completion = self->mCompletion;
+        dispatch_async(self->mQueue, ^{
+            completion(response, nil);
+        });
+        delete self;
+    }
+
+    Credentials::DeviceAttestationVerifier * mVerifier;
+    MTRDeviceAttestationInfo * mInfo;
+    dispatch_queue_t mQueue;
+    MTRDeviceAttestationVerificationHandler mCompletion;
+    Callback::Callback<Credentials::DeviceAttestationVerifier::OnAttestationInformationVerification> mCallback;
+};
+} // anonymous namespace
+
+- (void)verifyAttestationWithQueue:(dispatch_queue_t)queue completion:(MTRDeviceAttestationVerificationHandler)completion
+{
+    // TODO: Should we be using the caller queue for our intermediate work?
+    auto * basicInfo = [[MTRBaseClusterBasicInformation alloc] initWithDevice:self endpointID:@(kRootEndpointId) queue:queue];
+    auto * opCreds = [[MTRBaseClusterOperationalCredentials alloc] initWithDevice:self endpointID:@(kRootEndpointId) queue:queue];
+
+    __block NSNumber * vendorID;
+    __block NSNumber * productID;
+    __block NSData * dac;
+    __block NSData * pai;
+    __block NSData * attestationElements;
+    __block NSData * attestationSignature;
+    __block NSData * attestationChallenge;
+
+    NSMutableData * nonce = [NSMutableData dataWithLength:kAttestationNonceLength];
+    CHIP_ERROR err = Crypto::DRBG_get_bytes(static_cast<uint8_t *>(nonce.mutableBytes), nonce.length);
+    if (err != CHIP_NO_ERROR) {
+        dispatch_async(queue, ^{
+            completion(nil, [MTRError errorForCHIPErrorCode:err]);
+        });
+    }
+
+    // Sure would be nice to have await/async.  Try to avoid infinite completion
```

Comment:
Maybe a more conventional way to avoid deep nesting would be to have a helper class holding the state for your async operation, and a separate method per step that looks something like
```
- (void)fetchVendorID {
    [something doSomethingWithCompletion:^(...) {
        if (error) {
            [self handleError:error];
        } else {
            _vendorId = result;
            [self fetchProductID];
        }
    }];
}
```